pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3407 @@
+    __slots__ = ()
+
+    def is_potentially_variable(self):
+        return False
+
+
+class _LinearOperatorExpression(ExpressionBase):
+    """
+    An 'abstract' class that defines the polynomial degree for a simple
+    linear operator
+    """
+
+    __slots__ = ()
+
+    def _compute_polynomial_degree(self, result):
+        # NB: We can't use max() here because None (non-polynomial)
+        # overrides a numeric value (and max() just ignores it)
+        ans = 0
+        for x in result:
+            if x is None:
+                return None
+            elif ans < x:
+                ans = x
+        return ans
+
+
+class RangedExpression(_LinearOperatorExpression):
+    """
+    Ranged expressions, which define relations with a lower and upper bound::
+
+        x < y < z
+        x <= y <= z
+
+    Args:
+        args (tuple): child nodes
+        strict (tuple): flags that indicates whether the inequalities are strict
+    """
+
+    __slots__ = ('_strict',)
+    PRECEDENCE = 9
+
+    def __init__(self, args, strict):
+        super(RangedExpression,self).__init__(args)
+        self._strict = strict
+
+    def nargs(self):
+        return 3
+
+    def construct_node(self, args, memo):
+        return self.__class__(args, self._strict)
+
+    def __getstate__(self):
+        state = super(RangedExpression, self).__getstate__()
+        for i in RangedExpression.__slots__:
+            state[i] = getattr(self, i)
+        return state
+
+    def __nonzero__(self):
+        return bool(self())
+
+    __bool__ = __nonzero__
+
+    def is_relational(self):
+        return True
+
+    def _precedence(self):
+        return RangedExpression.PRECEDENCE
+
+    def _apply_operation(self, result):
+        _l, _b, _r = result
+        if not self._strict[0]:
+            if not self._strict[1]:
+                return _l <= _b and _b <= _r
+            else:
+                return _l <= _b and _b < _r
+        elif not self._strict[1]:
+            return _l < _b and _b <= _r
+        else:
+            return _l < _b and _b < _r
+
+    def _to_string(self, values, verbose, smap, compute_values):
+        return "{0}  {1}  {2}  {3}  {4}".format(values[0], '<' if self._strict[0] else '<=', values[1], '<' if self._strict[1] else '<=', values[2])
+
+    def is_constant(self):
+        return (self._args_[0].__class__ in native_numeric_types or self._args_[0].is_constant()) and \
+               (self._args_[1].__class__ in native_numeric_types or self._args_[1].is_constant()) and \
+               (self._args_[2].__class__ in native_numeric_types or self._args_[2].is_constant())
+
+    def is_potentially_variable(self):
+        return (self._args_[1].__class__ not in native_numeric_types and \
+                self._args_[1].is_potentially_variable()) or \
+               (self._args_[0].__class__ not in native_numeric_types and \
+                self._args_[0].is_potentially_variable()) or \
+               (self._args_[2]._
```

Comment:
This is a very specific class to only define the name "sum" and the precedence, and makes me wonder why we have it, especially given the comments above about needing to check for the derived classes anyways.